From b79cbaa954942f43cffbfffb26788b8958e7e1d1 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 26 Dec 2024 11:09:00 +0800
Subject: [PATCH 124/254] mtk: hostapd: add 11v MLD BPCC critical update
 support

1. re-define the BIT 7 in capability information
2. set BIT 7 in capability information if there is CU on non-TX BSS
3. trigger beacon set on TX bss when there is BPCC CU on non-TX BSS

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/beacon.c              |  7 +++++++
 src/ap/ieee802_11.c          | 11 +++++++++++
 src/common/ieee802_11_defs.h |  2 +-
 3 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index 571cddb6c..a0cc8a9fa 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -3464,6 +3464,7 @@ int ieee802_11_set_bss_critical_update(struct hostapd_data *hapd,
 				       enum bss_crit_update_event event)
 {
 	struct hostapd_data *h;
+	bool update_txbss_beacon = false;
 
 	if (!hapd->conf->mld_ap)
 		return 0;
@@ -3495,6 +3496,9 @@ int ieee802_11_set_bss_critical_update(struct hostapd_data *hapd,
 	case BSS_CRIT_UPDATE_EVENT_EHT_OPERATION:
 		hapd->eht_mld_bss_param_change += 1;
 		hapd->eht_mld_bss_critical_update = BSS_CRIT_UPDATE_SINGLE;
+
+		if (hapd != hostapd_mbssid_get_tx_bss(hapd))
+			update_txbss_beacon = true;
 		break;
 	case BSS_CRIT_UPDATE_EVENT_RECONFIG:
 	case BSS_CRIT_UPDATE_EVENT_ATTLM:
@@ -3517,6 +3521,9 @@ int ieee802_11_set_bss_critical_update(struct hostapd_data *hapd,
 		h->eht_mld_bss_critical_update = BSS_CRIT_UPDATE_FLAG;
 	}
 
+	if (update_txbss_beacon)
+		ieee802_11_set_beacon(hostapd_mbssid_get_tx_bss(hapd));
+
 	return 0;
 }
 
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index c4f322020..1ed314b0a 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -289,6 +289,17 @@ u16 hostapd_own_capab_info(struct hostapd_data *hapd)
 #ifdef CONFIG_IEEE80211BE
 	if (hapd->conf->mld_ap && hapd->eht_mld_bss_critical_update)
 		capab |= WLAN_CAPABILITY_PBCC;
+
+	if (hapd->iconf->mbssid && hapd == hostapd_mbssid_get_tx_bss(hapd)) {
+		for (i = 1; i < hapd->iface->num_bss; i++) {
+			h = hapd->iface->bss[i];
+			if (hostapd_is_mld_ap(h) &&
+			    h->eht_mld_bss_critical_update) {
+				capab |= WLAN_CAPABILITY_NON_TX_BSSID_CU;
+				break;
+			}
+		}
+	}
 #endif /* CONFIG_IEEE80211BE */
 
 
diff --git a/src/common/ieee802_11_defs.h b/src/common/ieee802_11_defs.h
index 3315d5bea..8f5fda911 100644
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -109,7 +109,7 @@
 #define WLAN_CAPABILITY_PRIVACY BIT(4)
 #define WLAN_CAPABILITY_SHORT_PREAMBLE BIT(5)
 #define WLAN_CAPABILITY_PBCC BIT(6)
-#define WLAN_CAPABILITY_CHANNEL_AGILITY BIT(7)
+#define WLAN_CAPABILITY_NON_TX_BSSID_CU BIT(7)
 #define WLAN_CAPABILITY_SPECTRUM_MGMT BIT(8)
 #define WLAN_CAPABILITY_QOS BIT(9)
 #define WLAN_CAPABILITY_SHORT_SLOT_TIME BIT(10)
-- 
2.45.2

